### PR TITLE
Suggestions

### PR DESCRIPTION
**Pull request by manuartero**
Hi Leire,

I've created a branch based on your current work to propose some suggestions and highlight a few things.
I hope you find this valuable.

Let me know if you liked the exercise :)

### CODE REVIEW

File: src/assets/index.ts
Comment (manuartero) [inline src/assets/index.ts:1]: so , just to point out that you can create these "menu" `index.ts` files even for a map of assets ! cleaning up your hook file

File: src/assets/index.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import ground from "./2/ground.png";
+import ice from "./2/ice.png";
+import normal from "./2/normal.png";
+import poison from "./2/poison.png";
+import psychic from "./2/psychic.png";
+import rock from "./2/rock.png";
+import steel from "./2/steel.png";
+import water from "./2/water.png";
+
+export const typeImageMap: Record<KnownPokemonType, string> = {
```
Comment (manuartero) [inline src/assets/index.ts:19]: tip: this typing is extremely useful in Typescript: 

```ts
map: Record<Key, Value>
```

in this case: 

```ts
map: Record<ASubsetOfStringThatMeansSomethingInMyProject, string>
```

where `KnownPokemonType` is 

```ts
type KnownPokemonType =
  | "bug"
  | "electric"
  | "fairy"
  | "fighting"
  | "fire"
  | "flying"
  | "grass"
  | "ground"
  | "ice"
  | "normal"
  | "poison"
  | "psychic"
  | "rock"
  | "steel"
  | "water";
```

File: src/components/pokedex.css [hunk 2]
```diff
@@ -181,41 +182,42 @@
...
 .favorite-pokemon-container {
-  display: flex;
-  flex-wrap: wrap;
-  align-items: flex-start;
-  margin-bottom: auto;
+  display: grid;
+  grid-template-columns: repeat(2, 1fr);
+  width: 100%;
+  height: 100%;
 }
```
Comment (manuartero) [inline src/components/pokedex.css:216]: this took me a little bit i did want to point you that there is a better solution for the **"3 rows of favorite pokemons, and each row has 2 items max"**

so, my recommendation here: everything related to "how to display in the screen" (in this case, this is a table of elements, 3 rows, 2 columns)

DO avoid Js , and try to solve that in pure CSS. 

In this case , the correct approach might be a Grid Layout

https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout

https://css-tricks.com/snippets/css/complete-guide-grid/

File: src/components/pokedex.tsx [hunk 3]
```diff
@@ -41,30 +50,42 @@ export function Pokedex() {
...
+              <div key={index} className="favorite-pokemon-wrapper">
+                {favoritePokemon[index] && (
+                  <img
+                    src={favoritePokemon[index].imageSrc}
+                    alt={favoritePokemon[index].name}
+                    className="favorite-pokemon-image"
+                  />
+                )}
+              </div>
+            ))}
```
Comment (manuartero) [inline src/components/pokedex.tsx:74]: second part for the grid layout thing: an interesting tool in js is to create  a "6 empty-elements array" (`Array(6).fill(null).map(() => {})`m in this case, we have an array of 6 favorite pokemons and what we do is to "fill" the next position of that array

File: src/hooks/debilidades.ts [hunk 4]
```diff
@@ -58,27 +19,31 @@ export const usePokemonWeaknesses = (pokemon: Pokemon | undefined) => {
   return weaknesses;
 };
 
-const PokemonDetails = async ({ pokemon }: { pokemon: Pokemon }) => {
```
Comment (manuartero) [inline src/hooks/debilidades.ts:61]: Minor thing: in JavaScript, capitalized function names usually carry special meaning. 

It's a **convention** where functions intended to be classes are named with a capital letter. 

If I read a function named `PokemonDetails()`, I'd expect one of the following:

- its a component meant to be renderer as `<PokemonDetails />`
- its a function meant to be used as a class like: `new PokemonDetails()`
- its a function and the original writer is telling me that is creating some kind of object and probably has something like: `PokemonDetails.create()`

***

TL;DR if it's a regular function name it lowercase xD

File: src/components/pokedex.tsx [hunk 5]
```diff
@@ -95,38 +116,48 @@ export function Pokedex() {
...
+            )}
+          >
+            <div className="weaknesses">
+              <strong>Weaknesses:</strong>
+            </div>
             {weaknesses.map((weakness) => (
               <React.Fragment key={weakness}>
-                <img className="imagenes"
+                <img
+                  className="imagenes"
```
Comment (manuartero) [inline src/components/pokedex.tsx:152]: no changes here, just the prettier

File: src/models.ts [hunk 6]
```diff
@@ -4,6 +4,23 @@ export type PokemonUri = {
...
+  | "flying"
+  | "grass"
+  | "ground"
+  | "ice"
+  | "normal"
+  | "poison"
+  | "psychic"
+  | "rock"
+  | "steel"
+  | "water";
```
Comment (manuartero) [inline src/models.ts:22]: TypeScript tip: It's incredibly useful to define specific types of strings rather than just any string ... like "ok, its a string, but not **any** string but one of these" 

you'll find this pattern is super valuable in Ts

File: src/models.ts [hunk 7]
```diff
@@ -27,21 +44,29 @@ export type PokemonType = {
   };
 };
 
-export const getPokemonTypeDamage = async (url: string): Promise<PokemonTypeDamage> => {
+const pokemonTypeDamageCache = new Map<string, PokemonTypeDamage>();
+
+export const fetchPokemonTypeDamage = async (
```
Comment (manuartero) [inline src/models.ts:49]: naming tip: name your API functions with `fetchBlaBluBli()`

File: src/models.ts [hunk 8]
```diff
@@ -27,21 +44,29 @@ export type PokemonType = {
...
   const response = await fetch(url);
-  const data = await response.json();
-  return data;
+  const pokemonTypeDamage = await response.json();
+  pokemonTypeDamageCache.set(url, pokemonTypeDamage);
+  return pokemonTypeDamage;
 };
 
-export interface PokemonTypeDamage {
+export type PokemonTypeDamage = {
```
Comment (manuartero) [inline src/models.ts:61]: really they are the same ... but my **personal** recommendation is to use `type` instead of `interface` 
there is a lot of posts about this ts-war. Really you can choose any side, anyway

https://dev.to/manuartero/my-take-on-interface-vs-type-hc8

> remember, it's called TypeScript, not InterfaceScript

File: src/models.ts [hunk 9]
```diff
@@ -27,21 +44,29 @@ export type PokemonType = {
...
+): Promise<PokemonTypeDamage> => {
+  if (pokemonTypeDamageCache.has(url)) {
+    return pokemonTypeDamageCache.get(url)!;
+  }
   const response = await fetch(url);
-  const data = await response.json();
-  return data;
+  const pokemonTypeDamage = await response.json();
+  pokemonTypeDamageCache.set(url, pokemonTypeDamage);
+  return pokemonTypeDamage;
```
Comment (manuartero) [inline src/models.ts:58]: feature proposal: memory cache so you can reduce HTTP calls: 
using this simple `Set`

Captura de pantalla 2024-03-21 a las 20 48 18

File: src/components/pokedex.css [hunk 10]
```diff
@@ -181,41 +182,42 @@
...
   top: -10px;
-  right: -200px; 
+  right: -190px;
   background: none;
   border: none;
   cursor: pointer;
+  border-radius: 50%;
+  padding: 10px;
   background-color: white;
 }
```
Comment (manuartero) [inline src/components/pokedex.css:192]: minor thing but i expect you prefer this change ? 

Captura de pantalla 2024-03-21 a las 21 03 25

File: src/components/pokedex.css [hunk 11]
```diff
@@ -181,41 +182,42 @@
 .heart-button {
   position: absolute;
```
Comment (manuartero) [inline src/components/pokedex.css:183]: ok this is super cracy 😅  Try to avoid this! 

instead you'd need to use flexbox for positioning these elements and the panels
 
https://css-tricks.com/snippets/css/a-guide-to-flexbox/

xD
